### PR TITLE
Sample using TransactionalSession with CosmosDB transactions

### PR DESCRIPTION
**Pull request by timbussmann**
TODO: 
- [ ] Link to this sample from the main page
- [ ] V7 version

### CODE REVIEW

File: samples/transactional-session/cosmosdb/Core_8/Backend/Backend.csproj [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+    <TargetFramework>net6.0</TargetFramework>
+  </PropertyGroup>
+
+	<ItemGroup>
+		<PackageReference Include="Microsoft.Extensions.Hosting.WindowsServices" Version="6.*" />
+		<PackageReference Include="Microsoft.Extensions.Logging.EventLog" Version="6.*" />
+		<PackageReference Include="NServiceBus.Extensions.Hosting" Version="2.0.0-beta.2" />
+		<PackageReference Include="NServiceBus.Persistence.CosmosDB" Version="2.0.0-beta.2.15" />
+		<PackageReference Include="NServiceBus.TransactionalSession" Version="2.0.0-beta.1" />
+		<PackageReference Include="NServiceBus.Persistence.NonDurable" Version="1.0.0-beta.2" />
```
Comment (lailabougria) [inline samples/transactional-session/cosmosdb/Core_8/Backend/Backend.csproj:14]: This one can be removed

File: samples/transactional-session/cosmosdb/Core_8/Frontend/Frontend.csproj [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+    <TargetFramework>net6.0</TargetFramework>
+    <LangVersion>7.3</LangVersion>
+  </PropertyGroup>
+  <ItemGroup>
+    <PackageReference Include="Microsoft.Extensions.Hosting.WindowsServices" Version="6.*" />
+    <PackageReference Include="Microsoft.Extensions.Logging.EventLog" Version="6.*" />
+    <PackageReference Include="NServiceBus.Extensions.Hosting" Version="2.0.0-beta.2" />
+    <PackageReference Include="NServiceBus.Persistence.CosmosDB" Version="2.0.0-beta.2.15" />
+    <PackageReference Include="NServiceBus.TransactionalSession" Version="2.0.0-beta.1" />
+    <PackageReference Include="NServiceBus.Persistence.NonDurable" Version="1.0.0-beta.2" />
```
Comment (lailabougria) [inline samples/transactional-session/cosmosdb/Core_8/Frontend/Frontend.csproj:13]: Can be removed

File: samples/transactional-session/cosmosdb/sample.md [hunk 3]
```diff
@@ -0,0 +1,69 @@
+---
+title: Using Transactional Session with CosmosDB
+reviewed: 2022-07-20
+component: Core
+related:
+- nservicebus/messaging/transactionalsession
+---
+
+This sample uses the transactional session package with the CosmosDB persistence package to achieve transactionally consistent database changes and message operations.
```
Comment (lailabougria) [inline samples/transactional-session/cosmosdb/sample.md:9]:
```suggestion
This sample uses the `NServiceBus.TransactionalSession`-package with the `NServiceBus.Persistence.CosmosDB`-package to achieve transactionally consistent database changes and message operations.
```

File: samples/transactional-session/cosmosdb/sample.md [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+downloadbutton
```
Comment (lailabougria) [inline samples/transactional-session/cosmosdb/sample.md:11]: Is this a thing? 😮

File: samples/transactional-session/cosmosdb/sample.md [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+## Overview
+
+The sample contains a frontend and a backend service sharing a CosmosDB database instance. When the transactional session on the frontend is commited, an `Order` document is created with the status `Received` and an `OrderReceived` event is published. The backend service subscribes to the event and loads the order document to update it's status to `Accepted`.
```
Comment (lailabougria) [inline samples/transactional-session/cosmosdb/sample.md:15]:
```suggestion
The sample contains a frontend and a backend service both accessing the same CosmosDB database instance. When the `TransactionalSession` on the frontend is committed, an `Order` document is created with the status `Received` and an `OrderReceived` event is published. The backend service subscribes to the event and loads the order document to update its status to `Accepted`.
```

File: samples/transactional-session/cosmosdb/sample.md [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+    ITransactionalSession-->>Frontend: 
+    deactivate ITransactionalSession
+    Queue->>Backend: Process(orderReceived)
+    activate Backend
+    Backend->>CosmosDB: Update(order)
+    Backend-->>Queue: 
+    deactivate Backend
+```
+
+The database and transport operations are executed atomically when the session is committed. If the session is aborted all database and transport operations are rolled back.
```
Comment (lailabougria) [inline samples/transactional-session/cosmosdb/sample.md:32]:
```suggestion
The database and transport operations are executed atomically when the session is committed. If the session is aborted, all database and transport operations are rolled back.
```

File: samples/transactional-session/cosmosdb/sample.md [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+
+## Prerequisties
+
+The sample is intended to be used with the [CosmosDB emulator](https://docs.microsoft.com/en-us/azure/cosmos-db/local-emulator?tabs=ssl-netstd21) to run locally. Alternatively, a connection string to an Azure CosmosDB instance can be provided.
+
```
Comment (lailabougria) [inline samples/transactional-session/cosmosdb/sample.md:37]:
```suggestion
```

File: samples/transactional-session/cosmosdb/sample.md [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+
+## Configuration
+
+### Frontend
+
+The `Frontend` service needs to enable the transactional storage feature in the endpoint configuration:
```
Comment (lailabougria) [inline samples/transactional-session/cosmosdb/sample.md:43]:
```suggestion
The `Frontend` service needs to enable the transactional session feature in the endpoint configuration:
```

File: samples/transactional-session/cosmosdb/sample.md [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+
+snippet: cosmos-txsession-frontend-config
+
+Note that the `Outbox` feature must be enabled for the transactional session feature to work.
```
Comment (lailabougria) [inline samples/transactional-session/cosmosdb/sample.md:47]: Should we elaborate more on this? Or do we leave the batched scenario exclusively for the docs?

File: samples/transactional-session/cosmosdb/sample.md [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+
+The endpoint is also configured to use CosmosDB persister with a default container and partition key path:
```
Comment (lailabougria) [inline samples/transactional-session/cosmosdb/sample.md:49]:
```suggestion
The endpoint is also configured to use the CosmosDB persistence with a default container and partition key path:
```

File: samples/transactional-session/cosmosdb/sample.md [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+
+snippet: cosmos-txsession-frontend-persistence
+
+### Backend
+
+The `Backend` service contains a message handler for the `OrderReceived` event. CosmosDB also needs to be configured for CosmosDB:
```
Comment (lailabougria) [inline samples/transactional-session/cosmosdb/sample.md:55]:
```suggestion
The `Backend` service contains a message handler for the `OrderReceived` event. The `Backend` endpoint also needs to be configured for CosmosDB:
```

File: samples/transactional-session/cosmosdb/sample.md [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+
+snippet: cosmos-txsession-backend-persistence
+
+Note that the backend needs a partition key mapping configured for the `OrderReceived` to determine what partition needs to be used to access the order.
```
Comment (lailabougria) [inline samples/transactional-session/cosmosdb/sample.md:59]: Can we link to the docs on partitioning here?

File: samples/transactional-session/cosmosdb/sample.md [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+
+## Running the sample
+
+Start the `Frontend` and `Backend` services.
```
Comment (lailabougria) [inline samples/transactional-session/cosmosdb/sample.md:63]:
```suggestion
Start the `Frontend` and `Backend` endpoints.
```

File: samples/transactional-session/cosmosdb/sample.md (hunk 4 above)
Comment (timbussmann) [reply samples/transactional-session/cosmosdb/sample.md:11]: apparently, that was copied from other sample doc pages 😄

File: samples/transactional-session/cosmosdb/sample.md (hunk 9 above)
Comment (timbussmann) [reply samples/transactional-session/cosmosdb/sample.md:47]: I think that's for other docs

File: samples/transactional-session/cosmosdb/sample.md (hunk 12 above)
Comment (timbussmann) [inline samples/transactional-session/cosmosdb/sample.md:59]:
```suggestion
Note that the backend needs a [partition key mapping](/persistence/cosmosdb/transactions.md#specifying-the-partitionkey-to-use-for-the-transaction) configured for the `OrderReceived` to determine what partition needs to be used to access the order.
```

**Review by tmasternak — APPROVED** [review]

**Review by SzymonPobiega — APPROVED** [review]

File: samples/transactional-session/cosmosdb/sample.md [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+
+## Configuration
+
+### Frontend
+
+The `Frontend` service needs to enable the transactional session feature in the endpoint configuration:
+
+snippet: cosmos-txsession-frontend-config
+
+Note that the `Outbox` feature must be enabled to achieve atomicity with the transactional session feature.
```
Comment (tmasternak) [inline samples/transactional-session/cosmosdb/sample.md:46]:
```suggestion
NOTE: The `Outbox` feature must be enabled to achieve atomicity with the transactional session feature.
```

**Pull request merged**